src/IECore/MeshAlgo.cpp

```diff
@@ -1,16 +1,238 @@
 #include <assert.h>
 #include <algorithm>
 
+#include "boost/mpl/and.hpp"
 #include "OpenEXR/ImathVec.h"
 
 #include "IECore/MeshAlgo.h"
+#include "IECore/FaceVaryingPromotionOp.h"
+#include "IECore/DespatchTypedData.h"
+#include "IECore/DespatchTypedData.h"
+#include "IECore/TypeTraits.h"
 
 using namespace IECore;
 using namespace Imath;
 
 namespace
 {
 
+template< typename T > struct IsArithmeticVectorTypedData : boost::mpl::and_<
+	IECore::TypeTraits::IsNumericBasedVectorTypedData<T>,
+	boost::mpl::not_< IECore::TypeTraits::IsBox<typename TypeTraits::VectorValueType<T>::type > >,
+	boost::mpl::not_< IECore::TypeTraits::IsQuat<typename TypeTraits::VectorValueType<T>::type > >
+> {};
+
+
+struct  AverageValueFromVector
+{
+	typedef DataPtr ReturnType;
+
+	template<typename From> ReturnType operator()( typename From::ConstPtr data )
+	{
+		const typename From::ValueType &src = data->readable();
+		if ( src.size() )
+		{
+			return new TypedData< typename From::ValueType::value_type >( std::accumulate( src.begin() + 1, src.end(), *src.begin() ) / src.size() );
+		}
+		return NULL;
+	}
+};
+
+
+struct MeshVertexToUniform
+{
+	typedef DataPtr ReturnType;
+
+	MeshVertexToUniform( const MeshPrimitive *mesh )	:	m_mesh( mesh )
+	{
+	}
+
+	template<typename From> ReturnType operator()( const From* data )
+	{
+		typename From::Ptr result = static_cast< From* >( Object::create( data->typeId() ).get() );
+		typename From::ValueType &trg = result->writable();
+		const typename From::ValueType &src = data->readable();
+
+		trg.reserve( m_mesh->numFaces() );
+
+		std::vector<int>::const_iterator vId = m_mesh->vertexIds()->readable().begin();
```

Comment:
The logic seems fine here, but I wonder if you're aware of the `PolygonIterator` and `PolygonVertexIterator` classes? I think they could be applied to abstract away all the indexing...something like this (untested)...

```
for( PolygonIterator fIt = m_mesh->faceBegin(), feIt = m_mesh->faceEnd(); fIt != feIt; ++fIt )
{
    typedef VertexIterator<From::ValueType::const_iterator> VIt;
    VIt vBegin = fIt->vertexBegin( src.begin() );
    VIt vEnd = fIt->vertexEnd( src.begin() );
    trg.push_back( std::accumulate( vBegin + 1, vEnd(), *vBegin ) / ( vEnd - vBegin ) );
}
```

Perhaps there's a subtlety that means this isn't possible, but if so, it'd be nice to address it, since the intention with those iterators was to simplify this sort of thing. I don't want to stand in the way of getting this code into production, so if you/@andrewkaufman think it's not worth the effort right now, perhaps we could add a "todo" comment to revisit? The more we make use of the other library components, the more we can be sure they're actually fit for purpose...